bmc/connection.go

```diff
@@ -28,67 +29,105 @@ type connectionProviders struct {
 // OpenConnectionFromInterfaces will try all opener interfaces and remove failed ones.
 // The reason failed ones need to be removed is so that when other methods are called (like powerstate)
 // implementations that have connections wont nil pointer error when their connection fails.
-func OpenConnectionFromInterfaces(ctx context.Context, timeout time.Duration, generic []interface{}) (opened []interface{}, metadata Metadata, err error) {
-	var metadataLocal Metadata
+func OpenConnectionFromInterfaces(ctx context.Context, timeout time.Duration, providers []interface{}) (opened []interface{}, _ Metadata, err error) {
+	var metadata Metadata
 
-	for _, elem := range generic {
-		// return immediately if the context is done/terminated/etc
-		select {
-		case <-ctx.Done():
-			err = multierror.Append(err, ctx.Err())
-			return nil, metadata, err
-		default:
-		}
-		// get the provider name for use in metadata
-		providerName := getProviderName(elem)
-		// now, try to open connections
+	// Return immediately if the context is done.
+	select {
+	case <-ctx.Done():
+		err = multierror.Append(err, ctx.Err())
+		return nil, metadata, err
+	default:
+	}
+
+	// Create a context with the specified timeout. This is done for backward compatibility but
+	// we should consider removing the timeout parameter alltogether given the context will
+	// container the timeout.
+	ctx, cancel := context.WithTimeout(ctx, timeout)
```

Comment:
this feels like it should be right next to `o.Open` and a new context created per provider. otherwise, each provider could affect the context of another.  This is one reason that the current implementation creates a new context the way it does.

Reply:
What effects do you think could propagate between them?

Reply:
@jacobweinstock